File: python/eventgraph/store.py
"""Store protocol and InMemory implementation."""

from __future__ import annotations

import threading
from dataclasses import dataclass
from typing import Protocol

from .errors import ChainIntegrityError, EventNotFoundError
from .event import Event, compute_hash, canonical_form, canonical_content_json
from .types import ActorID, ConversationID, EventID, EventType, Hash, Option


@dataclass(frozen=True, slots=True)
class ChainVerification:
    """Result of a chain integrity check."""

    valid: bool
    length: int


class Store(Protocol):
    """Event persistence — append-only, hash-chained."""

    def append(self, event: Event) -> Event: ...
    def get(self, event_id: EventID) -> Event: ...
    def head(self) -> Option[Event]: ...
    def count(self) -> int: ...
    def verify_chain(self) -> ChainVerification: ...
    def by_type(self, event_type: EventType, limit: int) -> list[Event]: ...
    def by_source(self, source: ActorID, limit: int) -> list[Event]: ...
    def by_conversation(self, conversation_id: ConversationID, limit: int) -> list[Event]: ...
    def ancestors(self, event_id: EventID, max_depth: int) -> list[Event]: ...
    def descendants(self, event_id: EventID, max_depth: int) -> list[Event]: ...
    def close(self) -> None: ...


class InMemoryStore:
    """Thread-safe in-memory event store."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._events: list[Event] = []
        self._index: dict[str, int] = {}  # event_id -> position

    def append(self, event: Event) -> Event:
        with self._lock:
            if self._events:
                last = self._events[-1]
                if event.prev_hash != last.hash:
                    raise ChainIntegrityError(
                        len(self._events),
                        f"prev_hash {event.prev_hash.value} != head hash {last.hash.value}",
                    )
            self._events.append(event)
            self._index[event.id.value] = len(self._events) - 1
            return event

    def get(self, event_id: EventID) -> Event:
        with self._lock:
            pos = self._index.get(event_id.value)
            if pos is None:
                raise EventNotFoundError(event_id.value)
            return self._events[pos]

    def head(self) -> Option[Event]:
        with self._lock:
            if not self._events:
                return Option.none()
            return Option.some(self._events[-1])

    def count(self) -> int:
        with self._lock:
            return len(self._events)

    def verify_chain(self) -> ChainVerification:
        with self._lock:
            for i in range(1, len(self._events)):
                expected_prev = self._events[i - 1].hash
                actual_prev = self._events[i].prev_hash
                if expected_prev != actual_prev:
                    return ChainVerification(valid=False, length=i)
            return ChainVerification(valid=True, length=len(self._events))

    def recent(self, limit: int) -> list[Event]:
        """Return the most recent events, newest first."""
        with self._lock:
            return list(reversed(self._events[-limit:]))

    def by_type(self, event_type: EventType, limit: int) -> list[Event]:
        """Return events of the given type, newest first, up to limit."""
        with self._lock:
            matches: list[Event] = []
            for event in reversed(self._events):
                if event.type.value == event_type.value:
                    matches.append(event)
                    if len(matches) >= limit:
                        break
            return matches

    def by_source(self, source: ActorID, limit: int) -> list[Event]:
        """Return events from the given source, newest first, up to limit."""
        with self._lock:
            matches: list[Event] = []
            for event in reversed(self._events):
                if event.source.value == source.value:
                    matches.append(event)
                    if len(matches) >= limit:
                        break
            return matches

    def by_conversation(self, conversation_id: ConversationID, limit: int) -> list[Event]:
        """Return events in the given conversation, newest first, up to limit."""
        with self._lock:
            matches: list[Event] = []
            for event in reversed(self._events):
                if event.conversation_id.value == conversation_id.value:
                    matches.append(event)
                    if len(matches) >= limit:
                        break
            return matches

    def ancestors(self, event_id: EventID, max_depth: int) -> list[Event]:
        """Return causal ancestors via BFS on causes, up to max_depth levels."""
        with self._lock:
            result: list[Event] = []
            seen: set[str] = {event_id.value}
            # Current frontier: causes of the starting event
            pos = self._index.get(event_id.value)
            if pos is None:
                raise EventNotFoundError(event_id.value)
            frontier: list[str] = [
                c.value for c in self._events[pos].causes
                if c.value != event_id.value  # skip bootstrap self-ref
            ]
            for _ in range(max_depth):
                next_frontier: list[str] = []
                for eid in frontier:
                    if eid in seen:
                        continue
                    seen.add(eid)
                    p = self._index.get(eid)
                    if p is None:
                        continue
                    evt = self._events[p]
                    result.append(evt)
                    for c in evt.causes:
                        if c.value not in seen:
                            next_frontier.append(c.value)
                frontier = next_frontier
                if not frontier:
                    break
            return result

    def descendants(self, event_id: EventID, max_depth: int) -> list[Event]:
        """Return causal descendants via BFS on reverse-cause index, up to max_depth levels."""
        with self._lock:
            if self._index.get(event_id.value) is None:
                raise EventNotFoundError(event_id.value)
            # Build reverse index lazily
            children: dict[str, list[str]] = {}
            for evt in self._events:
                for c in evt.causes:
                    if c.value != evt.id.value:  # skip bootstrap self-ref
                        children.setdefault(c.value, []).append(evt.id.value)

            result: list[Event] = []
            seen: set[str] = {event_id.value}
            frontier = children.get(event_id.value, [])
            for _ in range(max_depth):
                next_frontier: list[str] = []
                for eid in frontier:
                    if eid in seen:
                        continue
                    seen.add(eid)
                    p = self._index.get(eid)
                    if p is None:
                        continue
                    evt = self._events[p]
                    result.append(evt)
                    for child_id in children.get(eid, []):
                        if child_id not in seen:
                            next_frontier.append(child_id)
                frontier = next_frontier
                if not frontier:
                    break
            return result

    def close(self) -> None:
        pass

File: dotnet/src/EventGraph/Store.cs
namespace EventGraph;

/// <summary>Result of a chain integrity check.</summary>
public readonly record struct ChainVerification(bool Valid, int Length);

/// <summary>Event persistence protocol — append-only, hash-chained.</summary>
public interface IStore
{
    Event Append(Event ev);
    Event Get(EventId eventId);
    Option<Event> Head();
    int Count();
    ChainVerification VerifyChain();
    List<Event> ByType(EventType type, int limit);
    List<Event> BySource(ActorId source, int limit);
    List<Event> ByConversation(ConversationId id, int limit);
    List<Event> Ancestors(EventId id, int maxDepth);
    List<Event> Descendants(EventId id, int maxDepth);
    void Close();
}

/// <summary>Thread-safe in-memory event store.</summary>
public sealed class InMemoryStore : IStore
{
    private readonly Lock _lock = new();
    private readonly List<Event> _events = new();
    private readonly Dictionary<string, int> _index = new();

    public Event Append(Event ev)
    {
        lock (_lock)
        {
            if (_events.Count > 0)
            {
                var last = _events[^1];
                if (ev.PrevHash != last.Hash)
                    throw new ChainIntegrityException(
                        _events.Count,
                        $"prev_hash {ev.PrevHash.Value} != head hash {last.Hash.Value}");
            }
            _events.Add(ev);
            _index[ev.Id.Value] = _events.Count - 1;
            return ev;
        }
    }

    public Event Get(EventId eventId)
    {
        lock (_lock)
        {
            if (_index.TryGetValue(eventId.Value, out var pos))
                return _events[pos];
            throw new EventNotFoundException(eventId.Value);
        }
    }

    public Option<Event> Head()
    {
        lock (_lock)
        {
            return _events.Count == 0
                ? Option<Event>.None()
                : Option<Event>.Some(_events[^1]);
        }
    }

    public int Count()
    {
        lock (_lock) { return _events.Count; }
    }

    public ChainVerification VerifyChain()
    {
        lock (_lock)
        {
            for (int i = 1; i < _events.Count; i++)
            {
                if (_events[i - 1].Hash != _events[i].PrevHash)
                    return new ChainVerification(false, i);
            }
            return new ChainVerification(true, _events.Count);
        }
    }

    public List<Event> Recent(int limit)
    {
        lock (_lock)
        {
            var start = Math.Max(0, _events.Count - limit);
            var result = _events.GetRange(start, _events.Count - start);
            result.Reverse();
            return result;
        }
    }

    public List<Event> ByType(EventType type, int limit)
    {
        lock (_lock)
        {
            var result = new List<Event>();
            for (int i = _events.Count - 1; i >= 0 && result.Count < limit; i--)
            {
                if (_events[i].Type == type)
                    result.Add(_events[i]);
            }
            return result;
        }
    }

    public List<Event> BySource(ActorId source, int limit)
    {
        lock (_lock)
        {
            var result = new List<Event>();
            for (int i = _events.Count - 1; i >= 0 && result.Count < limit; i--)
            {
                if (_events[i].Source == source)
                    result.Add(_events[i]);
            }
            return result;
        }
    }

    public List<Event> ByConversation(ConversationId id, int limit)
    {
        lock (_lock)
        {
            var result = new List<Event>();
            for (int i = _events.Count - 1; i >= 0 && result.Count < limit; i--)
            {
                if (_events[i].ConversationId == id)
                    result.Add(_events[i]);
            }
            return result;
        }
    }

    public List<Event> Ancestors(EventId id, int maxDepth)
    {
        lock (_lock)
        {
            var result = new List<Event>();
            var visited = new HashSet<string>();
            var queue = new Queue<(EventId Id, int Depth)>();
            queue.Enqueue((id, 0));
            visited.Add(id.Value);

            while (queue.Count > 0)
            {
                var (currentId, depth) = queue.Dequeue();
                if (depth > maxDepth) continue;

                if (!_index.TryGetValue(currentId.Value, out var pos))
                    continue;

                var ev = _events[pos];
                if (depth > 0) // don't include the starting event
                    result.Add(ev);

                if (depth < maxDepth)
                {
                    foreach (var causeId in ev.Causes)
                    {
                        if (visited.Add(causeId.Value))
                            queue.Enqueue((causeId, depth + 1));
                    }
                }
            }
            return result;
        }
    }

    public List<Event> Descendants(EventId id, int maxDepth)
    {
        lock (_lock)
        {
            // Build reverse index lazily
            var childIndex = new Dictionary<string, List<int>>();
            for (int i = 0; i < _events.Count; i++)
            {
                foreach (var causeId in _events[i].Causes)
                {
                    if (!childIndex.TryGetValue(causeId.Value, out var list))
                    {
                        list = new List<int>();
                        childIndex[causeId.Value] = list;
                    }
                    list.Add(i);
                }
            }

            var result = new List<Event>();
            var visited = new HashSet<string>();
            var queue = new Queue<(string Id, int Depth)>();
            queue.Enqueue((id.Value, 0));
            visited.Add(id.Value);

            while (queue.Count > 0)
            {
                var (currentId, depth) = queue.Dequeue();
                if (depth > maxDepth) continue;

                if (depth > 0 && _index.TryGetValue(currentId, out var pos))
                    result.Add(_events[pos]);

                if (depth < maxDepth && childIndex.TryGetValue(currentId, out var children))
                {
                    foreach (var childPos in children)
                    {
                        var childId = _events[childPos].Id.Value;
                        if (visited.Add(childId))
                            queue.Enqueue((childId, depth + 1));
                    }
                }
            }
            return result;
        }
    }

    public void Close() { }
}
